### PR TITLE
PYTHON-2899 Remove code for MongoDB <= 3.4

### PR DESCRIPTION
**Pull request by ShaneHarvey**

### CODE REVIEW

**Review by juliusgeo — APPROVED** [review]
LGTM!

File: pymongo/auth.py [hunk 1]
```diff
@@ -496,10 +492,8 @@ def _authenticate_default(credentials, sock_info):
             return _authenticate_scram(credentials, sock_info, 'SCRAM-SHA-256')
         else:
             return _authenticate_scram(credentials, sock_info, 'SCRAM-SHA-1')
-    elif sock_info.max_wire_version >= 3:
-        return _authenticate_scram(credentials, sock_info, 'SCRAM-SHA-1')
     else:
-        return _authenticate_mongo_cr(credentials, sock_info)
```
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply pymongo/auth.py:502]: It is still supported by pymongo. What I've removed here is the default auth path for MongoDB 2.6 servers.

File: pymongo/pool.py [hunk 2]
```diff
@@ -781,13 +771,8 @@ def legacy_write(self, request_id, msg, max_doc_size, with_last_error):
           - `max_doc_size`: size in bytes of the largest document in `msg`.
           - `with_last_error`: True if a getlasterror command is appended.
         """
-        self._raise_if_not_writable(not with_last_error)
-
+        self._raise_if_not_writable(True)
         self.send_message(msg, max_doc_size)
```
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply pymongo/pool.py:772]: It's always OP_MSG, the docs were out of date. Updated.

File: pymongo/message.py [hunk 3]
```diff
@@ -1585,7 +1271,7 @@ def unpack(cls, msg):
         # PYTHON-945: ignore starting_from field.
         flags, cursor_id, _, number_returned = cls.UNPACK_FROM(msg)
 
-        documents = bytes(msg[20:])
+        documents = msg[20:]
```
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply pymongo/message.py:1258]: It's the memoryview created in network._receive_data_on_socket().

File: pymongo/message.py [hunk 4]
```diff
@@ -1665,71 +1351,11 @@ def unpack(cls, msg):
         if len(msg) != first_payload_size + 5:
             raise ProtocolError("Unsupported OP_MSG reply: >1 section")
 
-        payload_document = bytes(msg[5:])
+        payload_document = msg[5:]
```
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply pymongo/message.py:1338]: It's the memoryview created in network._receive_data_on_socket().

**Review by ShaneHarvey — COMMENTED** [review]
Updated to remove the unused OP_KILL_CURSORS code path.

**Review by behackett — APPROVED** [review]
LGTM! No going back now.

**Pull request merged**